Lib/test/test_glob.py

```diff
@@ -513,7 +513,14 @@ def fn(pat):
             return glob.translate(pat, recursive=True, include_hidden=True, seps=['/', '\\'])
         self.assertEqual(fn('foo/bar\\baz'), r'(?s:foo[/\\]bar[/\\]baz)\Z')
         self.assertEqual(fn('**/*'), r'(?s:(?:.+[/\\])?[^/\\]+)\Z')
-
+        self.assertEqual(fn('foo[!a]bar'), r'(?s:foo(?![/\\])[^a]bar)\Z')
+        self.assertEqual(fn('foo[%-0]bar'), r'(?s:foo(?![/\\])[%-0]bar)\Z')
+        self.assertEqual(fn('foo[%-0][1-9]bar'), r'(?s:foo(?![/\\])[%-0][1-9]bar)\Z')
+        self.assertEqual(fn('foo[0-%]bar'), r'(?s:foo(?!)bar)\Z')
+        self.assertEqual(fn('foo[^-'), r'(?s:foo\[\^\-)\Z')
```

Comment:
We need also a test case with multiple ranges and incomplete ones, e.g., `[0-%][0-%[0-%]`. And possibly with an additional tail after the last range.